src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java

```diff
@@ -0,0 +1,246 @@
+                emitNodeInserted(node.getStudy().getId(), node.getParentNode().getIdNode(), node.getIdNode());
+                return nodeInfo;
+            }).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        });
+    }
+
+    public Mono<Void> deleteNode(UUID id, boolean deleteChildren) {
+        return Mono.fromRunnable(() -> {
+            List<UUID> removedNodes = new ArrayList<>();
+            UUID studyId = getStudyUuidForNodeId(id);
+            deleteNodes(id, deleteChildren, false, removedNodes);
+            emitNodesDeleted(studyId, removedNodes, deleteChildren);
+        });
+    }
+
+    public UUID getStudyUuidForNodeId(UUID id) {
+        Optional<NodeEntity> node = nodesRepository.findById(id);
+        return node.orElseThrow().getStudy().getId();
+    }
+
+    private void deleteNodes(UUID id, boolean deleteChildren, boolean allowDeleteRoot, List<UUID> removedNodes) {
+        Optional<NodeEntity> optNodeToDelete = nodesRepository.findById(id);
+        optNodeToDelete.ifPresent(nodeToDelete -> {
+            /* root cannot be deleted by accident */
+            if (!allowDeleteRoot && nodeToDelete.getType() == NodeType.ROOT) {
+                throw new StudyException(CANT_DELETE_ROOT_NODE);
+            }
+            if (!deleteChildren) {
+                nodesRepository.findAllByParentNodeIdNode(id).forEach(node -> {
+                    node.setParentNode(nodeToDelete.getParentNode());
+                    nodesRepository.save(node);
+                });
+            } else {
+                nodesRepository.findAllByParentNodeIdNode(nodeToDelete.getIdNode())
+                    .forEach(child -> deleteNodes(child.getIdNode(), true, false, removedNodes));
+            }
+            removedNodes.add(nodeToDelete.getIdNode());
+            repositories.get(nodeToDelete.getType()).deleteByNodeId(id);
+            nodesRepository.delete(nodeToDelete);
+        });
+    }
+
+    @Transactional
+    public void deleteRoot(UUID studyId) {
+        try {
+            List<NodeEntity> nodes = nodesRepository.findAllByStudyId(studyId);
+            repositories.forEach((key, repository) ->
+                repository.deleteAll(
+                    nodes.stream().filter(n -> n.getType().equals(key)).map(NodeEntity::getIdNode).collect(Collectors.toSet()))
+            );
+            nodesRepository.deleteAll(nodes);
+        } catch (EntityNotFoundException ignored) {
+            // nothing to do
+        }
+    }
+
+    @Transactional
+    public void createRoot(StudyEntity study) {
+        NodeEntity node = nodesRepository.save(new NodeEntity(null, null, NodeType.ROOT, study));
+        var root = RootNode.builder().studyId(study.getId()).id(node.getIdNode()).name("Root").build();
+        repositories.get(node.getType()).createNodeInfo(root);
+    }
+
+    @Transactional
+    public Mono<RootNode> getStudyTree(UUID studyId) {
+        List<NodeEntity> nodes = nodesRepository.findAllByStudyId(studyId);
```

Comment:
to be coherent with the rest, we should do the database access after the subscription of the mono